### PR TITLE
[106225716] Add "without-<feature>" classes to document for all inactive features.

### PR DESCRIPTION
**Pull request by mpelzsherman**
This allows us to keep the original CSS separate from the feature CSS by enclosing the original CSS inside the "without-<feature>" class.

https://www.pivotaltracker.com/story/show/106225716

### CODE REVIEW

**Comment by lfender6445** [comment]
can you provide example use case for this scenario? is this for when 2 or more tests are simultaneously active?

**Comment by mpelzsherman** [comment]
In a nutshell, we are de-coupling the global and feature-flipped CSS styles.

This is useful whenever you are implementing any pinball-flippable feature.

Say you have a style in your current global stylesheet, e.g.:

```css
#header {
  height: 40px;
  background: red;
}
```

In your feature's css file, you want to make the header blue, so you do this:

```css
.use-my-feature {
  #header {
    background: blue;
  }
}
```

Pinball automatically adds the .use-my-feature class to the entire document when `my_feature` is active, so this works fine. However, when you go to *publish* `my_feature`, you have to merge your style customizations with the global stylesheet to pick up the `height` value which was previously being "inherited" from the global stylesheet. This process can be very difficult and error-prone, as we have experienced with previous feature-publishing attempts.

This PR will cause pinball to add a `without-my-feature` class to the document for all **inactive** features. This forces you to put all of the styling you need in your feature-specific CSS, and all your global styles under a `without-<feature>` class, so that when your feature goes live, you can simply replace the entire `without-<feature>` section with your feature CSS.

Note that this change is only in the `css_tagger`, so it is effectively optional. Projects not wishing to use this feature can opt out by simply not upgrading to this version of the `css_tagger`.

Hope this makes sense. Happy to discuss further.

**Comment by c0** [comment]
Thanks @mpelzsherman for elaborating. I wanted to echo the rationale —

With the current version, it's recommended to set default styles and then *override* them with `.use-{feature}`. Overriding CSS with more specific rules is generally error prone (hence the need for patterns such as Modular CSS, BEM, etc).

This PR puts the CSS rules in *parallel* so that only one is applied at a time.

This is typical with the current version of pinball:
```scss
#header {
  background: red;
}

.use-my-feature {
  // override with a more specific selector
  #header {
    background: blue;
  }
}
```

And this PR:

```scss
.without-my-feature {
  #header {
    background: red;
  }

.use-my-feature {
  // not overwritten
  #header {
    background: blue;
  }
}
```

File: dist/css_tagger.min.js [hunk 1]
```diff
@@ -1,3 +1 @@
-// Minified on http://closure-compiler.appspot.com/ using Advanced settings
```
Comment (c0) [inline dist/css_tagger.min.js]: Do you mind restoring this comment?

File: src/css_tagger.coffee [hunk 2]
```diff
@@ -19,19 +19,17 @@ define ->
     classNames = []
 
     add = (name) ->
-      classNames.push 'use-' + name.split('_').join('-')
+      classToAdd = 'use-' + name.split('_').join('-')
```
Comment (c0) [inline src/css_tagger.coffee]: Can we DRY up `name.split('_').join('-')`?

File: src/css_tagger.coffee [hunk 3]
```diff
@@ -41,12 +39,26 @@ define ->
...
+
+      switch entry[0]
+        when 'activate'
+          add entry[1]
+
+        when 'add'
+          for feature, state of entry[1]
+            if state == 'active'
+              add feature
+            else if state == 'inactive'
```
Comment (c0) [inline src/css_tagger.coffee]: What do you think about a generic `else`? That way it would add disabled features.

File: src/css_tagger.coffee [hunk 4]
```diff
@@ -19,19 +19,17 @@ define ->
...
     add = (name) ->
-      classNames.push 'use-' + name.split('_').join('-')
+      classToAdd = 'use-' + name.split('_').join('-')
+      if !added(name)
+        classNames.push classToAdd
 
-    # Activated by the queue
-    for entry in pinballQueue
-      continue unless entry.length
+    added = (name) ->
```
Comment (c0) [inline src/css_tagger.coffee]: Can we rename this to something like `isAdded`?

**Comment by c0** [comment]
Overall, this is great! A few suggestions:

- Can we add a specific test for `.without-{feature}`? That way it's clear what the intent is.
- I think it's safe to add it to the main `pinball_wizard.coffee` too. Pinball should be opinionated. If someone doesn't want to use `.without-{feature}`, then they don't have to write CSS rules.
- Update the README :D

After this is merged, please bump to `1.0.0` :boom:

File: dist/css_tagger.min.js (hunk 1 above)
Comment (mpelzsherman) [inline dist/css_tagger.min.js]: I didn't use that minifier. I used [uglifyjs](https://www.npmjs.com/package/uglify-js) with the `--mangle` option.

Is there some reason that one is preferred?

Comment (c0) [inline dist/css_tagger.min.js]: Not one over another. Sounds like `uglifyjs` is easier to work with. Do you mind documenting it somewhere?

Comment (mpelzsherman) [inline dist/css_tagger.min.js]: Not at all, although the ideal thing would be for the `gulp dist` task to create it automatically. Currently, `gulp clean` deletes the file. This should be addressed in a separate PR IMO.

Comment (c0) [inline dist/css_tagger.min.js]: :+1: Thanks for writing up #19

**Comment by c0** [comment]
:boom:
